File: legostore/mosaic.go
package legostore

import (
	"github.com/raydog99/kvs/legostore/dataretriever"
	"github.com/raydog99/kvs/legostore/kvstore"
	"github.com/raydog99/kvs/legostore/metadata"
	"github.com/raydog99/kvs/legostore/strategies"
	"strconv"
	"strings"
	"time"
)

type Mosaic struct {
	Metadata      *metadata.Metadata
	KVStore       *kvstore.KVStore
	DataRetriever *dataretriever.DataRetriever
}

func NewMosaic() *Mosaic {
	return &Mosaic{
		Metadata:      metadata.NewMetadata(),
		KVStore:       kvstore.NewKVStore(),
		DataRetriever: dataretriever.NewDataRetriever(),
	}
}

func (m *Mosaic) SetupDevices() {
	m.Metadata.AddDevice(metadata.Device{
		ID:          1,
		Mnt:         "/mnt/nvme",
		Name:        "NVMe SSD",
		Capacity:    1000,
		Threads:     8,
		Compression: metadata.ZSTD,
		CostPerGB:   0.25,
	})
	m.Metadata.AddDevice(metadata.Device{
		ID:          2,
		Mnt:         "/mnt/ssd",
		Name:        "SATA SSD",
		Capacity:    2000,
		Threads:     4,
		Compression: metadata.LZ4,
		CostPerGB:   0.15,
	})
	m.Metadata.AddDevice(metadata.Device{
		ID:          3,
		Mnt:         "/mnt/hdd",
		Name:        "HDD",
		Capacity:    4000,
		Threads:     1,
		Compression: metadata.None,
		CostPerGB:   0.05,
	})
}

func (m *Mosaic) SetupTables() {
	customerColumns := []metadata.Column{
		{ID: 1, TableID: 1, Name: "c_custkey", DataType: "INT", Size: 4 * 150000},
		{ID: 2, TableID: 1, Name: "c_name", DataType: "VARCHAR", Size: 25 * 150000},
		{ID: 3, TableID: 1, Name: "c_address", DataType: "VARCHAR", Size: 40 * 150000},
	}
	m.Metadata.AddTable(metadata.Table{ID: 1, Name: "customer", Columns: customerColumns})

	ordersColumns := []metadata.Column{
		{ID: 4, TableID: 2, Name: "o_orderkey", DataType: "INT", Size: 4 * 1500000},
		{ID: 5, TableID: 2, Name: "o_custkey", DataType: "INT", Size: 4 * 1500000},
		{ID: 6, TableID: 2, Name: "o_totalprice", DataType: "DECIMAL", Size: 8 * 1500000},
	}
	m.Metadata.AddTable(metadata.Table{ID: 2, Name: "orders", Columns: ordersColumns})
}

func (m *Mosaic) ParseQuery(queryStr string) metadata.TraceEntry {
	parts := strings.Split(queryStr, ":")
	tableID, _ := strconv.Atoi(parts[0])
	columnIDs := []uint32{}
	for _, colIDStr := range strings.Split(parts[1], ",") {
		colID, _ := strconv.Atoi(colIDStr)
		columnIDs = append(columnIDs, uint32(colID))
	}
	return metadata.TraceEntry{TableID: uint32(tableID), ColumnIDs: columnIDs}
}

func (m *Mosaic) RunQuery(queryStr string) {
	query := m.ParseQuery(queryStr)
	m.DataRetriever.FetchTableScan(query.TableID, query.ColumnIDs, m.Metadata, m.KVStore)
	result := m.DataRetriever.GetBufferData()
	println("Query executed:", queryStr)
	println("Retrieved", len(result), "columns")
}

func (m *Mosaic) OptimizeStorage(budget float32) {
	placements := strategies.OptimizePlacement(budget, m.Metadata)
	strategies.ApplyPlacements(placements, m.Metadata, m.KVStore, m.DataRetriever)
}

func (m *Mosaic) RunMosaic(budget float32, queries []string) {
	println("Initial benchmark...")
	initialTime := m.Benchmark(queries)
	println("Initial execution time:", initialTime, "s")

	println("\nOptimizing storage...")
	m.OptimizeStorage(budget)

	println("\nRunning benchmark after optimization...")
	optimizedTime := m.Benchmark(queries)
	println("Optimized execution time:", optimizedTime, "s")

	improvement := (initialTime - optimizedTime) / initialTime * 100
	println("\nImprovement:", improvement, "%")
}

func (m *Mosaic) Benchmark(queries []string) float64 {
	start := time.Now()
	for _, query := range queries {
		m.RunQuery(query)
	}
	elapsed := time.Since(start)
	return elapsed.Seconds()
}

File: epaxos/src/mosaic.rs
use crate::metadata::{Metadata, Device, Table, Column, TraceEntry, CompressionAlgorithm};
use crate::kv_store::KVStore;
use crate::data_retriever::DataRetriever;
use crate::strategies::{optimize_placement, apply_placements};

pub struct Mosaic {
    pub metadata: Metadata,
    pub kv_store: KVStore,
    pub data_retriever: DataRetriever,
}

impl Mosaic {
    pub fn new() -> Self {
        Mosaic {
            metadata: Metadata::new(),
            kv_store: KVStore::new(),
            data_retriever: DataRetriever::new(),
        }
    }

    pub fn setup_devices(&mut self) {
        self.metadata.add_device(Device {
            id: 1,
            mnt: "/mnt/nvme".to_string(),
            name: "NVMe SSD".to_string(),
            capacity: 1000,
            threads: 8,
            compression: CompressionAlgorithm::ZSTD,
            cost_per_gb: 0.25,
        });
        self.metadata.add_device(Device {
            id: 2,
            mnt: "/mnt/ssd".to_string(),
            name: "SATA SSD".to_string(),
            capacity: 2000,
            threads: 4,
            compression: CompressionAlgorithm::LZ4,
            cost_per_gb: 0.15,
        });
        self.metadata.add_device(Device {
            id: 3,
            mnt: "/mnt/hdd".to_string(),
            name: "HDD".to_string(),
            capacity: 4000,
            threads: 1,
            compression: CompressionAlgorithm::None,
            cost_per_gb: 0.05,
        });
    }

    pub fn setup_tables(&mut self) {
        let customer_columns = vec![
            Column { id: 1, table_id: 1, name: "c_custkey".to_string(), data_type: "INT".to_string(), size: 4 * 150000 },
            Column { id: 2, table_id: 1, name: "c_name".to_string(), data_type: "VARCHAR".to_string(), size: 25 * 150000 },
            Column { id: 3, table_id: 1, name: "c_address".to_string(), data_type: "VARCHAR".to_string(), size: 40 * 150000 },
        ];
        self.metadata.add_table(Table { id: 1, name: "customer".to_string(), columns: customer_columns });

        let orders_columns = vec![
            Column { id: 4, table_id: 2, name: "o_orderkey".to_string(), data_type: "INT".to_string(), size: 4 * 1500000 },
            Column { id: 5, table_id: 2, name: "o_custkey".to_string(), data_type: "INT".to_string(), size: 4 * 1500000 },
            Column { id: 6, table_id: 2, name: "o_totalprice".to_string(), data_type: "DECIMAL".to_string(), size: 8 * 1500000 },
        ];
        self.metadata.add_table(Table { id: 2, name: "orders".to_string(), columns: orders_columns });
    }

    pub fn parse_query(&self, query_str: &str) -> TraceEntry {
        let parts: Vec<&str> = query_str.split(':').collect();
        let table_id = parts[0].parse().unwrap();
        let column_ids: Vec<u32> = parts[1].split(',').map(|s| s.parse().unwrap()).collect();
        TraceEntry { table_id, column_ids }
    }

    pub fn run_query(&mut self, query_str: &str) {
        let query = self.parse_query(query_str);
        self.data_retriever.fetch_table_scan(query.table_id, &query.column_ids, &self.metadata, &self.kv_store);
        let result = self.data_retriever.get_buffer_data();
        println!("Query executed: {}", query_str);
        println!("Retrieved {} columns", result.len());
    }

    pub fn optimize_storage(&mut self, budget: f32) {
        let placements = optimize_placement(budget, &self.metadata);
        apply_placements(&placements, &self.metadata, &mut self.kv_store, &mut self.data_retriever);
    }

    pub fn run_mosaic(&mut self, budget: f32, queries: &[String]) {
        println!("Initial benchmark...");
        let initial_time = self.benchmark(queries);
        println!("Initial execution time: {:.2} s", initial_time);

        println!("\nOptimizing storage...");
        self.optimize_storage(budget);

        println!("\nRunning benchmark after optimization...");
        let optimized_time = self.benchmark(queries);
        println!("Optimized execution time: {:.2} s", optimized_time);

        let improvement = (initial_time - optimized_time) / initial_time * 100.0;
        println!("\nImprovement: {:.2}%", improvement);
    }

    pub fn benchmark(&mut self, queries: &[String]) -> f32 {
        use std::time::Instant;
        let start = Instant::now();
        for query in queries {
            self.run_query(query);
        }
        start.elapsed().as_secs_f32()
    }
}